cosmo_tester/test_suites/cluster/full_cluster_test.py

```diff
@@ -34,3 +41,165 @@ def test_cluster_single_db(cluster_with_single_db, logger, attributes, cfy):
                      cert_path=mgr2.local_ca)
 
     check_managers(mgr1, mgr2)
+
+
+def test_queue_node_failover(cluster_with_single_db, logger,
+                             module_tmpdir, attributes, ssh_key, cfy):
+    hello_world = prepare_cluster_with_agent(
+        cluster_with_single_db, logger,
+        module_tmpdir, attributes, ssh_key, cfy
+    )
+    broker1, broker2, broker3, db, mgr1, mgr2 = cluster_with_single_db
+    validate_cluster_and_agents(mgr1)
+    agent_broker_ip1 = verify_agent_broker_connection_and_get_broker_ip(mgr1)
+
+    # stop the rabbitmq service in the agent's broker node
+    agent_broker1 = None
+    for broker in [broker1, broker2, broker3]:
+        if broker.private_ip_address == agent_broker_ip1:
+            agent_broker1 = broker
+            break
+    with agent_broker1.ssh() as broker_ssh:
+        broker_ssh.run('sudo service cloudify-rabbitmq stop')
+
+    # the agent should now pick another broker
+    validate_cluster_and_agents(mgr1, expected_broker_status='Degraded')
+    agent_broker_ip2 = \
+        verify_agent_broker_connection_and_get_broker_ip(mgr1)
+    assert agent_broker_ip2 != agent_broker_ip1
+
+    """
+    The following asserts that the agent will reconnect to a stopped and
+    restarted broker.
+    """
+    restarted_broker_connected = False
+
+    # bring the stopped broker back up
+    with agent_broker1.ssh() as broker_ssh:
+        broker_ssh.run('sudo service cloudify-rabbitmq start')
+    wait_for_healthy_broker_cluster(mgr1)
+    # stop another broker
+    agent_broker2 = None
+    for broker in [broker1, broker2, broker3]:
+        if broker.private_ip_address == agent_broker_ip2:
+            agent_broker2 = broker
+            break
+    with agent_broker2.ssh() as broker_ssh:
+        broker_ssh.run('sudo service cloudify-rabbitmq stop')
+    agent_broker_ip3 = \
+        verify_agent_broker_connection_and_get_broker_ip(mgr1)
+
+    # if the agent connected to the initial broker, we're done
```

Comment:
This bit might be better written as: while not restart_broker_connected:

Reply:
You could put a limit on it since it should take at most 2 more, so assert at the end of the loop that you haven't hit attempt 3 or something similar (to avoid the test just 'hanging' forever).